In [ ]:
#import necessary libraries

import pandas as pd
import math as math

In [ ]:
#load in csv of all our data
Data = pd.read_csv(r'C:\Users\Brent\Desktop\ClassNotes\CSS844\Project1\VegetativeIndices\vegetation_indices_results_2.csv')

In [ ]:
Accumulated_GDD = Data['Accumulated.GDD'][:5] #makes new GDD series
Data = Data.drop('Accumulated.GDD', axis=1) #removes GDD from data
Data_VI = Data.drop('id', axis=1) #makes new Dataframe just for making New_Data
Data_VI = Data_VI.drop('Date', axis=1)

In [ ]:
#New empty panda to write data into
New_Data = pd.DataFrame()
New_Data['id'] = []
New_Data['id'] = Data['id'].unique() #Makes a row for each ID
for i in range(len(Data_VI.columns)): #Makes a column for each VI_Date
    for j in range(0,5):
        new_column = f"{Data_VI.columns[i]}_{Data['Date'][j]}"
        New_Data[new_column] = None
print(New_Data)

In [ ]:
#Reorganizing data into the new dataframe
for i in range(len(New_Data.columns)-1): #Iterates through each column
    temp_series = [] #sets up empty series
    for j in range(len(New_Data)): #iterates through each row of that column
        next_value = Data.loc[(j*5)+(i % 5)][math.floor(i/5)+2] #chooses the next value

        temp_series.append(next_value) #adds the value to the series
    New_Data[New_Data.columns[i+1]] = temp_series #replaces column with the series


New_Data.head()

id  NDVI_7.12  NDVI_8.08  NDVI_8.22  NDVI_8.31  NDVI_10.02  \
0  EL23_9001   0.167980   0.791446   0.838585   0.774085    0.493593   
1  EL23_9002   0.144880   0.797365   0.879939   0.781144    0.428886   
2  EL23_9003   0.177046   0.788489   0.890482   0.846157    0.656544   
3  EL23_9004   0.146934   0.775339   0.852731   0.807269    0.558960   
4  EL23_9005   0.136395   0.697297   0.839898   0.725110    0.613957   

   NDRE_7.12  NDRE_8.08  NDRE_8.22  NDRE_8.31  ...  SAVI_7.12  SAVI_8.08  \
0   0.068363   0.359697   0.517505   0.437886  ...   0.063754   0.075265   
1   0.055943   0.383222   0.593344   0.467933  ...   0.056148   0.073061   
2   0.072639   0.388713   0.608876   0.545172  ...   0.068785   0.076913   
3   0.056491   0.344146   0.555419   0.499668  ...   0.055520   0.076336   
4   0.048857   0.210170   0.442532   0.356881  ...   0.051894   0.057812   

   SAVI_8.22  SAVI_8.31  SAVI_10.02  MSAVI_7.12  MSAVI_8.08  MSAVI_8.22  \
0   0.632611   0.553385    0.296957    0.049553    0.053177    0.677161   
1   0.630094   0.573528    0.233504    0.043671    0.051518    0.684690   
2   0.664864   0.617288    0.430437    0.053721    0.054418    0.733304   
3   0.643102   0.587021    0.344481    0.043020    0.054015    0.693661   
4   0.583337   0.524744    0.390614    0.040206    0.040423    0.614291   

   MSAVI_8.31  MSAVI_10.02  
0    0.569041     0.268355  
1    0.594881     0.202939  
2    0.659872     0.415125  
3    0.614835     0.318078  
4    0.530919     0.369057  

[5 rows x 36 columns]

In [158]:
#Looking at averages of each VI at each date
Averages = New_Data.iloc[:, 1:].mean()
display(Averages)
display(Accumulated_GDD)

NDVI_7.12          0.137214
NDVI_8.08          0.611884
NDVI_8.22          0.771410
NDVI_8.31          0.687606
NDVI_10.02         0.668242
NDRE_7.12          0.057616
NDRE_8.08          0.228781
NDRE_8.22          0.463868
NDRE_8.31          0.374675
NDRE_10.02         0.296925
EVI_7.12           0.041515
EVI_8.08           0.041582
EVI_8.22           0.553105
EVI_8.31           0.498086
EVI_10.02          0.470126
CI_7.12            0.812974
CI_8.08            2.053768
CI_8.22            4.041195
CI_8.31            4.585833
CI_10.02           3.796053
RTVIcore_7.12      0.517844
RTVIcore_8.08      0.686608
RTVIcore_8.22     19.557290
RTVIcore_8.31     17.023666
RTVIcore_10.02    13.376063
SAVI_7.12          0.046019
SAVI_8.08          0.047845
SAVI_8.22          0.509740
SAVI_8.31          0.478383
SAVI_10.02         0.457407
MSAVI_7.12         0.035045
MSAVI_8.08         0.033331
MSAVI_8.22         0.517269
MSAVI_8.31         0.478161
MSAVI_10.02        0.450950
dtype: float64

0     498.418750
1    1039.087500
2    1283.100000
3    1433.216667
4    1868.450000
Name: Accumulated.GDD, dtype: float64

Accumulated GDDS:  
7.12 - 498.418750  
8.08 - 1039.087500  
8.22 - 1283.100000  
8.31 - 1433.216667  
10.02 - 1868.450000  

Maximums:  
NDVI_8.22  
NDRE_8.22  
EVI_8.22  
CI_8.31  
RTVIcore_8.22 - Throw out?  
SAVI_8.22  
MSAVI_8.22  

In [160]:
Final_Data = New_Data #Make a new dataframe to work in

#Calculate VI changes per GDD from 7.12 to Max Date
Final_Data['NDVI_per_GDD_Max'] = (Final_Data['NDVI_8.22'] - Final_Data['NDVI_7.12']) / (Accumulated_GDD[2]-Accumulated_GDD[0])
Final_Data['NDRE_per_GDD_Max'] = (Final_Data['NDRE_8.22'] - Final_Data['NDRE_7.12']) / (Accumulated_GDD[2]-Accumulated_GDD[0])
Final_Data['EVI_per_GDD_Max'] = (Final_Data['EVI_8.22'] - Final_Data['EVI_7.12']) / (Accumulated_GDD[2]-Accumulated_GDD[0])
Final_Data['CI_per_GDD_Max'] = (Final_Data['CI_8.31'] - Final_Data['CI_7.12']) / (Accumulated_GDD[3]-Accumulated_GDD[0])
Final_Data['RTVIcore_per_GDD_Max'] = (Final_Data['RTVIcore_8.22'] - Final_Data['RTVIcore_7.12']) / (Accumulated_GDD[2]-Accumulated_GDD[0])
Final_Data['SAVI_per_GDD_Max'] = (Final_Data['SAVI_8.22'] - Final_Data['SAVI_7.12']) / (Accumulated_GDD[2]-Accumulated_GDD[0])
Final_Data['MSAVI_per_GDD_Max'] = (Final_Data['MSAVI_8.22'] - Final_Data['MSAVI_7.12']) / (Accumulated_GDD[2]-Accumulated_GDD[0])

#Calculate VI changes per GDD from Max Date to 10.02
Final_Data['NDVI_per_GDD_End'] = (Final_Data['NDVI_10.02'] - Final_Data['NDVI_8.22']) / (Accumulated_GDD[4]-Accumulated_GDD[2])
Final_Data['NDRE_per_GDD_End'] = (Final_Data['NDRE_10.02'] - Final_Data['NDRE_8.22']) / (Accumulated_GDD[4]-Accumulated_GDD[2])
Final_Data['EVI_per_GDD_End'] = (Final_Data['EVI_10.02'] - Final_Data['EVI_8.22']) / (Accumulated_GDD[4]-Accumulated_GDD[2])
Final_Data['CI_per_GDD_End'] = (Final_Data['CI_10.02'] - Final_Data['CI_8.31']) / (Accumulated_GDD[4]-Accumulated_GDD[2])
Final_Data['RTVIcore_per_GDD_End'] = (Final_Data['RTVIcore_10.02'] - Final_Data['RTVIcore_8.22']) / (Accumulated_GDD[4]-Accumulated_GDD[2])
Final_Data['SAVI_per_GDD_End'] = (Final_Data['SAVI_10.02'] - Final_Data['SAVI_8.22']) / (Accumulated_GDD[4]-Accumulated_GDD[2])
Final_Data['MSAVI_per_GDD_End'] = (Final_Data['MSAVI_10.02'] - Final_Data['MSAVI_8.22']) / (Accumulated_GDD[4]-Accumulated_GDD[2])

#Calcuate VI changes per GDD from 7.12 to 10.02
Final_Data['NDVI_per_GDD_Total'] = (Final_Data['NDVI_10.02'] - Final_Data['NDVI_7.12']) / (Accumulated_GDD[4]-Accumulated_GDD[0])
Final_Data['NDRE_per_GDD_Total'] = (Final_Data['NDRE_10.02'] - Final_Data['NDRE_7.12']) / (Accumulated_GDD[4]-Accumulated_GDD[0])
Final_Data['EVI_per_GDD_Total'] = (Final_Data['EVI_10.02'] - Final_Data['EVI_7.12']) / (Accumulated_GDD[4]-Accumulated_GDD[0])
Final_Data['CI_per_GDD_Total'] = (Final_Data['CI_10.02'] - Final_Data['CI_7.12']) / (Accumulated_GDD[4]-Accumulated_GDD[0])
Final_Data['RTVIcore_per_GDD_Total'] = (Final_Data['RTVIcore_10.02'] - Final_Data['RTVIcore_7.12']) / (Accumulated_GDD[4]-Accumulated_GDD[0])
Final_Data['SAVI_per_GDD_Total'] = (Final_Data['SAVI_10.02'] - Final_Data['SAVI_7.12']) / (Accumulated_GDD[4]-Accumulated_GDD[0])
Final_Data['MSAVI_per_GDD_Total'] = (Final_Data['MSAVI_10.02'] - Final_Data['MSAVI_7.12']) / (Accumulated_GDD[4]-Accumulated_GDD[0])

Final_Data.head()

id  NDVI_7.12  NDVI_8.08  NDVI_8.22  NDVI_8.31  NDVI_10.02  \
0  EL23_9001   0.167980   0.791446   0.838585   0.774085    0.493593   
1  EL23_9002   0.144880   0.797365   0.879939   0.781144    0.428886   
2  EL23_9003   0.177046   0.788489   0.890482   0.846157    0.656544   
3  EL23_9004   0.146934   0.775339   0.852731   0.807269    0.558960   
4  EL23_9005   0.136395   0.697297   0.839898   0.725110    0.613957   

   NDRE_7.12  NDRE_8.08  NDRE_8.22  NDRE_8.31  ...  RTVIcore_per_GDD_End  \
0   0.068363   0.359697   0.517505   0.437886  ...             -0.034831   
1   0.055943   0.383222   0.593344   0.467933  ...             -0.037870   
2   0.072639   0.388713   0.608876   0.545172  ...             -0.031746   
3   0.056491   0.344146   0.555419   0.499668  ...             -0.033306   
4   0.048857   0.210170   0.442532   0.356881  ...             -0.017308   

   SAVI_per_GDD_End  MSAVI_per_GDD_End  NDVI_per_GDD_Total  \
0         -0.000573          -0.000698            0.000238   
1         -0.000678          -0.000823            0.000207   
2         -0.000400          -0.000544            0.000350   
3         -0.000510          -0.000642            0.000301   
4         -0.000329          -0.000419            0.000349   

   NDRE_per_GDD_Total  EVI_per_GDD_Total  CI_per_GDD_Total  \
0            0.000119           0.000160          0.001479   
1            0.000125           0.000118          0.001371   
2            0.000177           0.000267          0.002598   
3            0.000168           0.000204          0.001879   
4            0.000177           0.000243          0.002193   

   RTVIcore_per_GDD_Total  SAVI_per_GDD_Total  MSAVI_per_GDD_Total  
0                0.004902            0.000170             0.000160  
1                0.003989            0.000129             0.000116  
2                0.008674            0.000264             0.000264  
3                0.006880            0.000211             0.000201  
4                0.007747            0.000247             0.000240  

[5 rows x 57 columns]

In [161]:
#Write to CSV

Final_Data.to_csv('GAPITPhenotypes.csv', index=False, sep=',', na_rep='NA')